clone.go

```diff
@@ -0,0 +1,262 @@
+- Are there multiple possible outputs of our ligation reaction? For example, ligations may be
+  to create a "library" of plasmids, in which there are millions of valid combinations.
+- How do we handle sequences that get ligated in multiple orientations?
+
+These cloning functions handle all those problems so that they appear simple to the end user.
+
+In particular, there is a focus here on GoldenGate Assembly:
+https://en.wikipedia.org/wiki/Golden_Gate_Cloning
+https://www.neb.com/applications/cloning-and-synthetic-biology/dna-assembly-and-cloning/golden-gate-assembly
+
+GoldenGate is a particular kind of restriction enzyme cloning reaction that you can do
+in a single tube and that is extraordinarily efficient (up to 50 parts) and is popular
+for new modular DNA part toolkits. Users can easily simulate GoldenGate assembly reactions
+with just their input fragments + the enzyme name.
+
+Let's build some DNA!
+
+Keoni
+
+******************************************************************************/
+
+// CloneSequence is a simple struct that can carry a circular or linear DNA sequence.
+type CloneSequence struct {
+	Sequence string
+	Circular bool
+}
+
+// Overhang is a struct that represents the ends of a linearized sequence where Enzymes had cut.
+type Overhang struct {
+	Length   int
+	Position int
+	Forward  bool
+}
+
+// Fragment is a struct that represents linear DNA sequences with sticky ends.
+type Fragment struct {
+	Sequence        string
+	ForwardOverhang string
+	ReverseOverhang string
+}
+
+// Enzyme is a struct that represents restriction enzymes.
+type Enzyme struct {
+	EnzymeName        string
+	RegexpFor         *regexp.Regexp
+	RegexpRev         *regexp.Regexp
+	EnzymeSkip        int
+	EnzymeOverhangLen int
+	Directional       bool
+}
+
+/******************************************************************************
+
+Base cloning functions begin here.
+
+******************************************************************************/
+
+// https://qvault.io/2019/10/21/golang-constant-maps-slices/
+func getBaseRestrictionEnzymes() map[string]Enzyme {
+	// Eventually, we want to get the data for this map from ftp://ftp.neb.com/pub/rebase
+	enzymeMap := make(map[string]Enzyme)
+
+	// Build default enzymes
+	enzymeMap["BsaI"] = Enzyme{"BsaI", regexp.MustCompile("GGTCTC"), regexp.MustCompile("GAGACC"), 1, 4, true}
+	enzymeMap["BbsI"] = Enzyme{"BbsI", regexp.MustCompile("GAAGAC"), regexp.MustCompile("GTCTTC"), 2, 4, true}
+	enzymeMap["BtgZI"] = Enzyme{"BtgZI", regexp.MustCompile("GCGATG"), regexp.MustCompile("CATCGC"), 10, 4, true}
+	enzymeMap["BsmBI"] = Enzyme{"BsmBI", regexp.MustCompile("CGTCTC"), regexp.MustCompile("GAGACG"), 1, 4, true}
+
+	// Return EnzymeMap
+	return enzymeMap
+}
+
+// RestrictionEnzymeCut cuts a given sequence with an enzyme represented by the enzyme's name.
+func RestrictionEnzymeCut(seq CloneSequence, enzymeStr string) ([]Fragment, error) {
```

Comment:
```suggestion
// a convenience wrapper around CutSequenceWithEnzyme that allows us to specify the enzyme by name.
func CutSequenceWithEnzymeByName(seq CloneSequence, enzymeStr string) ([]Fragment, error) {
```

Also test for this as well please!